lib/THC/generic/THCStorage.c

```diff
@@ -17,18 +17,23 @@ int THCStorage_(elementSize)(THCState *state)
   return sizeof(real);
 }
 
-void THCStorage_(set)(THCState *state, THCStorage *self, long index, real value)
+void THCStorage_(set)(THCState *state, THCStorage *self, long index, hostreal _value)
 {
   THArgCheck((index >= 0) && (index < self->size), 2, "index out of bounds");
+  real value = hostrealToReal(_value);
   THCudaCheck(cudaMemcpy(self->data + index, &value, sizeof(real), cudaMemcpyHostToDevice));
 }
 
-real THCStorage_(get)(THCState *state, const THCStorage *self, long index)
+hostreal THCStorage_(get)(THCState *state, const THCStorage *self, long index)
 {
   real value;
   THArgCheck((index >= 0) && (index < self->size), 2, "index out of bounds");
+#ifndef THC_REAL_IS_HALF
   THCudaCheck(cudaMemcpy(&value, self->data + index, sizeof(real), cudaMemcpyDeviceToHost));
-  return value;
+#else
+  THError("Get is not supported for CudaHalfStorage. Please convert to float first.");
```

Comment:
I'll try to implement this fp16->fp32 host conversion soon. I can't think of any other way of doing that quickly right now (launching a kernel seems too expensive for such a simple operation).